psi4/src/psi4/cc/ccenergy/local.cc

```diff
@@ -38,255 +38,1077 @@
-                C_DGEMM('n', 't', nvir, nvir, local_.pairdom_len[ij], 1.0, &(X2[0][0]), nso, &(local_.V[ij][0][0]),
-                        local_.pairdom_len[ij], 0.0, &(T2->matrix[0][ij][0]), nvir);
-            } else /* This must be a neglected weak pair; force it to zero */
-                memset((void *)T2->matrix[0][ij], 0, sizeof(double) * nvir * nvir);
-        }
-    }
-
-    free_block(X1);
-    free_block(X2);
-    free_block(T2tilde);
-    free_block(T2bar);
+    for (int ij = 0; ij < npairs; ++ij) {
+        if(!weak_pairs[ij]) {
+            int npno = survivors_list[ij];
+            auto T2temp = std::make_shared<Matrix>(nvir, nvir);
+            auto atemp = std::make_shared<Matrix>(nvir, npno);
+            auto btemp = std::make_shared<Matrix>(npno, npno);
+            auto T2bar = std::make_shared<Matrix>(npno, npno);
+
+            // If Q is nullptr, set T to 0
+            // and skip the iteration
+            if (npno == 0) { 
+                for(int ab=0; ab < nvir*nvir; ++ab) {
+                    int a = ab / nvir;
+                    int b = ab % nvir;
+                    T2->matrix[0][ij][ab] = 0.0;
+                }
+                continue; 
+            }
+
+            for(int ab=0; ab < nvir*nvir; ++ab) {
+                int a = ab / nvir;
+                int b = ab % nvir;
+                T2temp->set(a, b, T2->matrix[0][ij][ab]);
+            }
+
+            /* Transform the virtuals to the redundant projected virtual basis */
+            atemp->gemm(0, 0, 1, T2temp, Q[ij], 0);
+            T2bar->gemm(1, 0, 1, Q[ij], atemp, 0);
+
+            /* Transform the virtuals to the non-redundant virtual basis */
+            btemp->gemm(0, 0, 1, T2bar, L[ij], 0);
+            T2bar->gemm(1, 0, 1, L[ij], btemp, 0);
+
+            /* Divide the new amplitudes by the denominators */
+            int i = ij / nocc;
+            int j = ij % nocc;
+            for (int a = 0; a < npno; a++) {
+                for (int b = 0; b < npno; b++) {
+                    double val = T2bar->get(a,b);
+                    T2bar->set(a, b, val * 1.0 /
+                        (occ_eps[i] + occ_eps[j] - eps_pno[ij]->get(a) - eps_pno[ij]->get(b)));
+                }
+            }
+            /* Transform the new T2's to the redundant virtual basis */
+            btemp->gemm(0, 0, 1, L[ij], T2bar, 0);
+            T2bar->gemm(0, 1, 1, btemp, L[ij], 0);
+
+            /* Transform the new T2's to the MO basis */
+            atemp->gemm(0, 0, 1, Q[ij], T2bar, 0);
+            T2temp->gemm(0, 1, 1, atemp, Q[ij], 0);
+
+            for(int ab=0; ab < nvir*nvir; ++ab) {
+                int a = ab / nvir;
+                int b = ab % nvir;
+                T2->matrix[0][ij][ab] = T2temp->get(a,b);
+            }
+        }
+        else {
+            // Setting weak pair T2s to 0
+            for(int ab=0; ab < nvir*nvir; ++ab) {
+                T2->matrix[0][ij][ab] = 0.0;
```

Comment:
Surely there's a way to set the entire column to zero, rather than elementwise. `memset`, perhaps?